tests/conftest.py

```diff
@@ -2,49 +2,83 @@
 
 import os
 import sys
+import tempfile
 
 sys.path.append(os.path.join(os.path.dirname(__file__), "helpers"))
 
 import pytest
+
 from testplan import TestplanMock
-from testplan.common.utils import path
+from testplan.common.utils.path import default_runpath, makedirs
+
+
+# Testplan and various drivers have a `runpath` attribute in their config
+# and intermediate files will be placed under that path during running.
+# In PyTest we can invoke fixtures such as `runpath`, `runpath_module` and
+# `mockplan` so that a temporary `runpath` is generated. In order to easily
+# collect the output an environment variable `TEST_ROOT_RUNPATH` can be set,
+# it will be the parent directory of all those temporary `runpath`.
+
+
+def _generate_runpath():
+    """
+    Generate a temporary directory unless specified by environment variable.
+    """
+    if os.environ.get("TEST_ROOT_RUNPATH"):
+        yield tempfile.mkdtemp(dir=os.environ["TEST_ROOT_RUNPATH"])
+    else:
+        parent_runpath = default_runpath()
+        makedirs(parent_runpath)
+        # The path will be automatically removed after the test
+        with tempfile.TemporaryDirectory(dir=parent_runpath) as runpath:
+            yield runpath
+
+
+# For `runpath` series fixtures, We were originally using a pytest builtin
+# fixture called `tmp_path`, which will create a path in a form like:
+# "/tmp/pytest-of-userid/pytest-151/test_sub_pub_unsub0", but it has a known
+# issue: https://github.com/pytest-dev/pytest/issues/5456
 
 
 @pytest.fixture(scope="function")
 def runpath():
     """
-    Return a temporary runpath for testing. The path will be automatically
-    removed after the test.
-
-    We were originally using pytest builtin fixture tmp_path, which will create
-    a path in a form like: /tmp/pytest-of-userid/pytest-151/test_sub_pub_unsub0
-    But it has a known issue: https://github.com/pytest-dev/pytest/issues/5456
+    Return a temporary runpath for testing (function level).
     """
-    with path.TemporaryDirectory() as runpath:
-        yield runpath
+    yield from _generate_runpath()
 
 
 @pytest.fixture(scope="class")
 def runpath_class():
-    with path.TemporaryDirectory() as runpath:
-        yield runpath
+    """
+    Return a temporary runpath for testing (class level).
+    """
+    yield from _generate_runpath()
 
 
 @pytest.fixture(scope="module")
 def runpath_module():
-    with path.TemporaryDirectory() as runpath:
-        yield runpath
+    """
+    Return a temporary runpath for testing (module level).
+    """
+    yield from _generate_runpath()
 
 
 @pytest.fixture(scope="function")
 def mockplan():
     """
-    Return a temporary TestplanMock for testing. Some components needs a
-    testplan for getting runpath and cfg.
+    Return a temporary TestplanMock for testing. Some components need a
```

Comment:
Shall mockplan just take a runpath fixture as input, rather than repeating the logic?

Reply:
sure.